ocs_ci/utility/utils.py

```diff
@@ -397,27 +399,43 @@ def run_cmd(cmd, secrets=None, timeout=600, **kwargs):
         (str) Decoded stdout of command
 
     """
+    full_output = io.StringIO()
     masked_cmd = mask_secrets(cmd, secrets)
     log.info(f"Executing command: {masked_cmd}")
     if isinstance(cmd, str):
         cmd = shlex.split(cmd)
-    r = subprocess.run(
+    start_time = timeit.default_timer()
+    with subprocess.Popen(
         cmd,
         stdout=subprocess.PIPE,
         stderr=subprocess.PIPE,
-        stdin=subprocess.PIPE,
-        timeout=timeout,
+        universal_newlines=True,
         **kwargs
-    )
-    log.debug(f"Command output: {r.stdout.decode()}")
-    if r.stderr and not r.returncode:
-        log.warning(f"Command warning: {mask_secrets(r.stderr.decode(), secrets)}")
-    if r.returncode:
-        raise CommandFailed(
-            f"Error during execution of command: {masked_cmd}."
-            f"\nError is {mask_secrets(r.stderr.decode(), secrets)}"
-        )
-    return mask_secrets(r.stdout.decode(), secrets)
+    ) as p:
+        while True:
+            now = timeit.default_timer()
+            if (now - start_time) > timeout:
+                raise subprocess.TimeoutExpired(masked_cmd, timeout)
+            reads = [p.stdout.fileno(), p.stderr.fileno()]
+            ret = select.select(reads, [], [])
+            for fd in ret[0]:
+                if fd == p.stdout.fileno():
+                    read = p.stdout.readline()
+                    full_output.write(read)
+                    log.info(f"Command output: {read}")
+                if fd == p.stderr.fileno():
+                    read = p.stderr.readline()
+                    full_output.write(read)
+                    log.warning(f"Command warning: {mask_secrets(read, secrets)}")
```

Comment:
Is it necessary for us to have "Command output:" and "Command warning:" in these log messages? I feel like simply logging the message at info/warning level is enough without having to preface it with these strings.

I know this is carried over from what `run_cmd` has been doing but I'm bringing it up since we are here looking at this code.

Reply:
You have raised a very good point here. I've tried to keep the the the original functionality as it was before this attempt to implement a live output. If we agree to redirect both stdout and stderr through a single pipe, it will be much easier to implement and will require a smaller change in the code.

https://docs.python.org/3/library/subprocess.html#subprocess.CompletedProcess.stdout

Reply:
That sounds great, however the only thing I am curious about is the log level. Right now we are logging at different levels for stdout and stderr. If they are combined, I imagine this becomes more difficult to do.